keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/KLDivergence.java

```diff
@@ -0,0 +1,24 @@
+package io.improbable.keanu.algorithms.variational;
+
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.vertices.ProposalVertex;
+
+public class KLDivergence {
+    public static double compute(ProposalVertex q, NetworkSamples samples) {
+        double divergence = 0;
+
+        for (int i = 0; i < samples.size(); i++) {
+            double pLogProb = samples.getLogOfMasterP(i);
+            double qLogProb = q.getLogOfMasterP(samples.getSample(i));
+
+            if (pLogProb != Double.NEGATIVE_INFINITY) {
+                if (qLogProb == Double.NEGATIVE_INFINITY) {
+                    throw new IllegalArgumentException("Q cannot have smaller support than P");
```

Comment:
You should capture this in a test case. Also the converse, i.e. it's OK for P to have a smaller support than Q.